src/virtual_memory/shm.cpp

```diff
@@ -0,0 +1,104 @@
+/**
+ * @file
+ * @brief This file implements the virtual memory and virtual address handling
+ * @copyright Eta Scale AB. Licensed under the Eta Scale Open Source License. See the LICENSE file for details.
+ */
+
+#include <cerrno>
+#include <cstddef>
+#include <cstdlib>
+#include <fcntl.h>
+#include <iostream>
+#include <string>
+#include <sys/mman.h>
+#include <sys/stat.h>
+#include <sys/statvfs.h>
+#include <sys/types.h>
+#include <system_error>
+#include <unistd.h>
+#include "virtual_memory.hpp"
+
+namespace {
+	/* file constants */
+	/** @todo hardcoded start address */
+	const char* ARGO_START = (char*) 0x200000000000l;
+	/** @todo hardcoded end address */
+	const char* ARGO_END   = (char*) 0x600000000000l;
+	/** @todo hardcoded size */
+	const ptrdiff_t ARGO_SIZE = ARGO_END - ARGO_START;
+
+	/** @brief error message string */
+	const std::string msg_alloc_fail = "ArgoDSM could not allocate mappable memory";
+	/** @brief error message string */
+	const std::string msg_mmap_fail = "ArgoDSM failed to map in virtual address space.";
+	/** @brief error message string */
+	const std::string msg_main_mmap_fail = "ArgoDSM failed to set up virtual memory. Please report a bug.";
+
+	/* file variables */
+	/** @brief a file descriptor for backing the virtual address space used by ArgoDSM */
+	int fd;
+	/** @brief the address at which the virtual address space used by ArgoDSM starts */
+	void* start_addr;
+	/** @brief the size of the ArgoDSM virtual address space */
+	std::size_t avail;
+}
+
+namespace argo {
+	namespace virtual_memory {
+		void init() {
+			/* find maximum filesize */
+			struct statvfs b;
+			statvfs("/dev/shm", &b);
+			avail = b.f_bavail * b.f_bsize;
+			if(avail > static_cast<unsigned long>(ARGO_SIZE)/2u)
+				avail = ARGO_SIZE/2;
+			std::string filename = "/argocache" + std::to_string(getpid());
+			fd = shm_open(filename.c_str(), O_RDWR|O_CREAT, 0644);
+			if(shm_unlink(filename.c_str())) {
+				std::cerr << msg_main_mmap_fail << std::endl;
+				throw std::system_error(std::make_error_code(static_cast<std::errc>(errno)), msg_main_mmap_fail);
+				exit(EXIT_FAILURE);
+			}
+			if(ftruncate(fd, avail)) {
+				std::cerr << msg_main_mmap_fail << std::endl;
+				throw std::system_error(std::make_error_code(static_cast<std::errc>(errno)), msg_main_mmap_fail);
+				exit(EXIT_FAILURE);
+			}
+			/** @todo check desired range is free */
+			constexpr int flags = MAP_ANONYMOUS|MAP_SHARED|MAP_FIXED;
+			start_addr = ::mmap((void*)ARGO_START, avail, PROT_NONE, flags, -1, 0);
+			if(start_addr == MAP_FAILED) {
+				std::cerr << msg_main_mmap_fail << std::endl;
+				throw std::system_error(std::make_error_code(static_cast<std::errc>(errno)), msg_main_mmap_fail);
+				exit(EXIT_FAILURE);
+			}
+		}
+
+		void* start_address() {
+			return start_addr;
+		}
+
+		std::size_t size() {
+			return avail;
+		}
+
+		void* allocate_mappable(std::size_t alignment, std::size_t size) {
+			auto p = aligned_alloc(alignment, size);
```

Comment:
virtual_memory/shm.cpp:86:42: error: ‘aligned_alloc’ was not declared in this scope - Not in C++11. Confirmed to work with memalign instead.